src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -254,28 +321,31 @@ public static CallSite enumSwitch(MethodHandles.Lookup lookup,
         labels = labels.clone();
 
         Class<?> enumClass = invocationType.parameterType(0);
-        Stream.of(labels).forEach(l -> validateEnumLabel(enumClass, l));
-        MethodHandle temporary =
-                MethodHandles.insertArguments(DO_ENUM_SWITCH, 2, labels, lookup, enumClass, new ResolvedEnumLabels());
-        temporary = temporary.asType(invocationType);
+        labels = Stream.of(labels).map(l -> convertEnumConstants(lookup, enumClass, l)).toArray();
 
-        return new ConstantCallSite(temporary);
-    }
+        MethodHandle target;
+        boolean constantsOnly = Stream.of(labels).allMatch(l -> enumClass.isAssignableFrom(EnumDesc.class));
 
-    private static <E extends Enum<E>> void validateEnumLabel(Class<?> enumClassTemplate, Object label) {
-        if (label == null) {
-            throw new IllegalArgumentException("null label found");
-        }
-        Class<?> labelClass = label.getClass();
-        if (labelClass == Class.class) {
-            if (label != enumClassTemplate) {
-                throw new IllegalArgumentException("the Class label: " + label +
-                                                   ", expected the provided enum class: " + enumClassTemplate);
-            }
-        } else if (labelClass != String.class) {
-            throw new IllegalArgumentException("label with illegal type found: " + labelClass +
-                                               ", expected label of type either String or Class");
+        if (labels.length > 0 && constantsOnly) {
+            //If all labels are enum constants, construct an optimized handle for repeat index 0:
+            //if (selector == null) return -1
+            //else if (idx == 0) return mappingArray[selector.ordinal()]; //mapping array created lazily
+            //else return "createRepeatIndexSwitch(labels)"
+            MethodHandle body =
+                    MethodHandles.guardWithTest(MethodHandles.dropArguments(NULL_CHECK, 0, int.class),
+                                                MethodHandles.dropArguments(MethodHandles.constant(int.class, -1), 0, int.class, Object.class),
+                                                MethodHandles.guardWithTest(MethodHandles.dropArguments(IS_ZERO, 1, Object.class),
+                                                                            createRepeatIndexSwitch(lookup, labels),
+                                                                            MethodHandles.insertArguments(MAPPED_ENUM_LOOKUP, 1, lookup, enumClass, labels, new EnumMap())));
+            target = MethodHandles.permuteArguments(body, MethodType.methodType(int.class, Object.class, int.class), 1, 0);
```

Comment:
This `guardWithTest` does the opposite to what’s described in the above comment.